hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+            ProtocolMessages.GET_TIPS: self.handle_get_tips,
+            ProtocolMessages.TIPS: self.handle_tips,
+            ProtocolMessages.TIPS_END: self.handle_tips_end,
+            # XXX: overriding ReadyState.handle_error
+            ProtocolMessages.ERROR: self.handle_error,
+            ProtocolMessages.GET_DATA: self.handle_get_data,
+            ProtocolMessages.DATA: self.handle_data,
+            ProtocolMessages.RELAY: self.handle_relay,
+            ProtocolMessages.NOT_FOUND: self.handle_not_found,
+        }
+
+    def handle_not_found(self, payload: str) -> None:
+        """ Handle a received NOT-FOUND message.
+        """
+        # XXX: NOT_FOUND is a valid message, but we shouldn't ever receive it unless the other peer is running with a
+        #                modified code or if there is a bug
+        self.log.warn('not found? close connection', payload=payload)
+        self.protocol.send_error_and_close_connection('Unexpected NOT_FOUND')
+
+    def handle_error(self, payload: str) -> None:
+        """ Override protocols original handle_error so we can recover a sync in progress.
+        """
+        assert self.protocol.connections is not None
+        # forward message to overloaded handle_error:
+        self.protocol.handle_error(payload)
+
+    def update_synced(self, synced: bool) -> None:
+        self._synced = synced
+
+    @inlineCallbacks
+    def run_sync(self) -> Generator[Any, Any, None]:
+        if not self._is_enabled:
+            self.log.debug('sync is disabled')
+            return
+        if self._is_running:
+            # Already running...
+            self.log.debug('already running')
+            return
+        self._is_running = True
+        try:
+            yield self._run_sync()
+        finally:
+            self._is_running = False
+
+    @inlineCallbacks
+    def _run_sync(self) -> Generator[Any, Any, None]:
+        """Run sync. This is the entrypoint for the sync.
+        It is always safe to call this method.
+        """
+
+        if self.receiving_stream:
+            # If we're receiving a stream, wait for it to finish before running sync.
+            # If we're sending a stream, do the sync to update the peer's synced block
+            self.log.debug('receiving stream, try again later')
+            return
+
+        if self.mempool_manager.is_running():
+            # It's running a mempool sync, so we wait until it finishes
+            self.log.debug('running mempool sync, try again later')
+            return
+
+        bestblock = self.tx_storage.get_best_block()
+        meta = bestblock.get_metadata()
+
+        self.log.debug('run sync', height=meta.height)
+
+        assert self.protocol.connections is not None
+        assert self.tx_storage.indexes is not None
+        assert self.tx_storage.indexes.deps is not None
+
+        if self.tx_storage.indexes.deps.has_needed_tx():
```

Comment:
At first I wasn't getting this logic, because I thought that `has_needed_tx` means that "we have all tx that we need", not that "some txs we need are missing", which is the opposite. I had to look into the class for more information.

Maybe we could add a comment here explaining that this is the case where we need to download missing txs, or change the name of this method.

Reply:
Good point. Maybe rename it to `has_pending_tx_dependencies()`?

Reply:
Since we're talking about names, in the beginning I was a bit confused by having to think of "reverse dependencies". I think "dependents" is a simpler and straightforward name that works in this context. If A is a dependency of B, than B is dependent on A.

I'm just throwing the suggestion here, although my opinion is not very strong on the need of changing this, as you just get used to reverse dependencies after a while.